### ISSUE TITLE
Unable to load application with shotgun

### ISSUE THREAD

**Issue by frodsan** [puma/puma#919]
I'm try to run an application in development using [shotgun](https://github.com/rtomayko/shotgun) and puma.

With 2.16.0, it works:

```
shotgun
== Shotgun/Puma on http://127.0.0.1:9393/
Puma 2.16.0 starting...
* Min threads: 0, max threads: 16
* Environment: development
* Listening on tcp://127.0.0.1:9393
```

With 3.0.2, it stopped working:

```
== Shotgun/Puma on http://127.0.0.1:9393/
[1149] Puma starting in cluster mode...
[1149] * Version 3.0.2 (ruby 2.3.0-p0), codename: Plethora of Penguin Pinatas
[1149] * Min threads: 5, max threads: 5
[1149] * Environment: development
[1149] * Process workers: 2
[1149] * Preloading application
[1149] * Listening on tcp://127.0.0.1:9393
[1149] Use Ctrl-C to stop
config/puma.rb:16:in `block in _load_from': uninitialized constant #<Class:#<Puma::DSL:0x007fe971ad0270>>::DB (NameError)
```

I think it's trying to load `config/puma.rb` by default. Here is the config file:

```ruby
# frozen_string_literal: true

workers_count = Integer(ENV["PUMA_WORKERS"] || 2)
threads_count = Integer(ENV["PUMA_THREADS"] || 5)

workers workers_count
threads threads_count, threads_count

rackup DefaultRackup
port ENV["PORT"]
environment ENV["RACK_ENV"]

preload_app!

before_fork do
  DB.disconnect
end
```

Any suggestions? Thanks.

**Comment by evanphx** [comment]
Ok, I have figured out the issue. It was added in 3.0 for puma to load a default config file when launched via rackup. Which means that your case worked in 2.16 because that puma config file was actually never loaded at all! The issue is actually with how Shotgun loads your app and how that interacts with puma. Shotgun appears to load the app in a special non-global constant scope (probably so it can reload it safely), and thus puma can never seen those constantly.

So basically using any hooks that need to reference your app's code won't work. That's not a puma limitation, that's how Shotgun intends for it to work.
